pkg/daemon/bgp/bgp.go

```diff
@@ -0,0 +1,353 @@
+	}
+
+	peeringLink, err := netlink.LinkByName(peeringInterfaceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get bgp peering link %v: %v", peeringInterfaceName, err)
+	}
+
+	existLinkAddress, err := containernetwork.ListAllAddress(peeringLink)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get link address for bgp peering interface %v: %v", peeringInterfaceName, err)
+	}
+
+	switch len(existLinkAddress) {
+	case 0:
+		return nil, fmt.Errorf("there is no valid address on bpg peering interface")
+	case 1:
+		manager.routerID = existLinkAddress[0].IP.String()
+		if existLinkAddress[0].IP.To4() == nil {
+			manager.routerV6Address = existLinkAddress[0].IP
+		} else {
+			manager.routerV4Address = existLinkAddress[0].IP
+		}
+	case 2:
+		if (existLinkAddress[0].IP.To4() != nil && existLinkAddress[1].IP.To4() == nil) ||
+			(existLinkAddress[0].IP.To4() == nil && existLinkAddress[1].IP.To4() != nil) {
+			for _, addr := range existLinkAddress {
+				if addr.IP.To4() == nil {
+					manager.routerV6Address = addr.IP
+				}
+				manager.routerV4Address = addr.IP
+			}
+
+			// Use v4 address as routerID by default if v4/v6 addresses exist at the same time.
+			manager.routerID = manager.routerV4Address.String()
+			break
+		}
+		fallthrough
+	default:
+		defaultV4Route, err := containernetwork.GetDefaultRoute(netlink.FAMILY_V4)
+		if err != nil && err != daemonutils.NotExist {
+			return nil, fmt.Errorf("failed to get v4 default route: %v", err)
+		}
+
+		defaultV6Route, err := containernetwork.GetDefaultRoute(netlink.FAMILY_V6)
+		if err != nil && err != daemonutils.NotExist {
+			return nil, fmt.Errorf("failed to get v6 default route: %v", err)
+		}
+
+		for _, addr := range existLinkAddress {
+			if defaultV4Route != nil {
+				if addr.IP.Equal(defaultV4Route.Src) || addr.IPNet.Contains(defaultV4Route.Gw) {
+					manager.routerV4Address = addr.IP
+				}
+			}
+
+			if defaultV6Route != nil {
+				if addr.IP.Equal(defaultV6Route.Src) || addr.IPNet.Contains(defaultV6Route.Gw) {
+					manager.routerV6Address = addr.IP
+				}
+			}
+		}
+
+		if manager.routerV4Address == nil && manager.routerV6Address == nil {
+			return nil, fmt.Errorf("failed to find valid address for bgp router")
+		}
+
+		if manager.routerV4Address != nil {
+			// Use v4 address as routerID by default if v4/v6 addresses exist at the same time.
+			manager.routerID = manager.routerV4Address.String()
+		} else {
+			manager.routerID = manager.routerV6Address.String()
+		}
+	}
+
+	go manager.bgpServer.Serve()
+	return manager, nil
+}
+
+func (m *Manager) RecordPeer(address, password string, asn int, gracefulRestartTime int32) {
+	if gracefulRestartTime == 0 {
+		gracefulRestartTime = 300
+	}
+
+	m.peerMap[address] = &peerInfo{
+		address:                address,
+		asn:                    asn,
+		gracefulRestartSeconds: uint32(gracefulRestartTime),
+		password:               password,
+	}
+}
+
+func (m *Manager) RecordSubnet(cidr *net.IPNet) {
```

Comment:
need mutex here?

Reply:
That's not needed, for all of the daemon addr/route/bgp... managers, will never be called concurrently.